Add semi-linear support

smartin-vossloh opened the pull request:
This short series adds support for semi-linear merge strategy.

To do so, the PR must be fast-forward and a merge-commit must be created for the merge.
The fast-forward action is leveraged to run the check when merging, and enhanced by adding 
the `--no-ff` option to the `git merge` command when configured so.

Closes #36

smartin-vossloh committed:
fast-forward: Add --no-ff merge support

This change allows to choose the merge strategy to use for merging
the pull-request.

A merge-strategy parameter is added, allowing to choose the no-ff way
(i.e. introducing a merge commit). The default strategy remains
fast-forwarding the branch.

Fixes https://github.com/sequoia-pgp/fast-forward/issues/36

smartin-vossloh committed:
README.md: Add section about semi-linear history

smartin-vossloh committed:
README.md: Add a note about changing the merge trigger

nwalfield commented:
Thanks for working on it!  I want to let you know that I've seen it but I won't have time to look at it until after FOSDEM.  If I don't respond by Feb 7, please feel free to ping me!

smartin-vossloh committed:
README.md: Clean-up double-spaces

smartin-vossloh committed:
fast-forward: Add option for no-ff merge commit generation

This change introduce a workflow setting (input) allowing to chose the
kind of generated merge commit.

This setting offers similar choices as those available under the GitHub
*Allow merge commit* option.

On pull-request events, the JSON keys set by the GitHub *Allow merge
commit* are available through the GitHub API, but not on PR comment
event. Hence, we have to set it in the workflow itself instead of
relying on the repository settings from GitHub.

smartin-vossloh committed:
README.md: Add section about the generated merge commit message

smartin-vossloh commented:
@nwalfield Thanks for the notice :) Meanwhile, I will run a bit more tests on it until get back on the PR.

Enjoy FOSDEM! :+1: (I won't be there this year, but plan to check the videos.)

smartin-vossloh commented:
Tests and semi-linear history example available here: https://github.com/smartin-vossloh/fast-forward/network

smartin-vossloh commented:
@nwalfield I'd really like to see this integrate and available in the fast-forward action. Hope you'll find sometime to review this PR.

smartin-vossloh commented:
Hi @nwalfield :slightly_smiling_face:

Do you have any news about this PR?
(whether it is wanted for the fast-forward action? if so do you have an idea when you will be available for reviewing this development?)

I'm looking forward to get your feedback.
TIA

smartin-vossloh commented:
@nwalfield up! 🙂

nwalfield commented on action.yml:
```diff
@@ -23,6 +23,13 @@ inputs:
+      When merging using the merge-commit method, set the merge commit
+      message content.
+
+      Possible values are: default, pr-title or pr-title-and-body.
```
The description should describe what the three options do.

nwalfield commented on src/fast-forward.sh:
```diff
@@ -302,15 +302,41 @@ LOG=$(mktemp)
+                MESSAGE=$(mktemp)
+                PR_NUMBER="$(github_pull_request .number)"
+                PR_HTML_URL="$(github_pull_request .html_url)"
+                # debug stuff
```
What does this comment mean?

nwalfield commented on src/fast-forward.sh:
```diff
@@ -302,15 +302,41 @@ LOG=$(mktemp)
+                            echo "* ${PR_HTML_URL}"
+                            echo "* from ${PR_REF} into ${BASE_REF}"
+                            ;;
+                        *)
```
This isn't great.  This should be an explicitly named option or default.  If the value doesn't match any of the known values, we should emit an error.

nwalfield commented on README.md:
```diff
@@ -1,7 +1,7 @@
 
 This repository contains a GitHub action that merges a pull request by
-fast forwarding the target branch.  The action is triggered when an
+fast forwarding the target branch. The action is triggered when an
```
I'm a bit confused about this and other similar changes.  It seems that this is doing the opposite of what the commit message says: instead of fixing the double spaces after full stops, it is removing them!

nwalfield reviewed (changes requested):
First, my apologies for the long time it took me to review these changes.

Most importantly, this change is missing a unit test.  The unit tests are run from [this action](https://github.com/sequoia-pgp/fast-forward/blob/main/.github/workflows/ci.yml), which calls [this script](https://github.com/sequoia-pgp/fast-forward/blob/main/tests/fast-forward.sh).  It would be good to add a second script that checks that a valid semi-linear merge works, and that an invalid one (one that can't be fast forwarded) is rejected.

The option handling should be improved.  An invalid value should always be rejected, and `default` should be an alias for the default value.

I appreciate that the two main changes are in separate commits.  I'd prefer that the documentation related to a commit be added in the same commit.

Finally, the stylistic change is unrelated to this set of changes and should be proposed in a separate MR.  That said, I'm not interested in that change.  I prefer having two spaces after a full stop, and I don't want the churn.

smartin-vossloh commented:
Hi @nwalfield ,

First, apologies for the delay.
I am still willing to get this dev in, but priority are not here right now.

I hope I will be able to rework following your review in the coming weeks.

Thanks for your patience.

nwalfield commented:
No worries.
